### ISSUE TITLE
acsetup.cfg contains no game name, settings write to "Saved Games/<game name>/"

### ISSUE THREAD

**Issue by morganwillcock** [adventuregamestudio/ags#520]
I was making a replacement launcher, but it seems this isn't possible using just the data from the config file.
The config file doesn't include any identifier for the game, but then writes config changes into a path which includes the game name. This makes it impossible to use another tool to edit the files.

As a side issue, but related to the above, winsetup is writing config data into the "Saved Games" directory of the user profile. Surely this should be going into roaming or local appdata? This isn't save game data and isn't affected by configuring a custom save game path.

**Comment by ivan-mogilko** [comment]
I was not sure where to store these settings and decided to place them in save folder to avoid creating any new paths (custom paths did not exist at that time). The location of the savegame folder is currently determined by the engine based on the game data, but it has to pre-read that folder setting from the game package first before proceeding with the game or running setup dialog.

This of course may be changed if there is a better alternative.

Speaking of config file, I was quite reluctant to rely on its contents in general, because it may be easily edited, this is why I made such decision to read most important properties from the game data instead. Maybe that was another case of overthinking the problem, but I could not figure out what should engine/setup do if this field is missing in config.

Since we are at this, in the past we were discussing how could alternate (non-embedded) setup program get some data from the game or engine (for example, engine's version etc), and we thought that passing some data by standard output may be not a bad idea. For example, if engine is called with something like "--tell gfxdrivers" it will return the list of supported renderers, and so forth. I just thought I'd mention this, since this is also another option to get paths or similar data from the game using engine as a tool to extract it.

**Comment by morganwillcock** [comment]
> Speaking of config file, I was quite reluctant to rely on its contents in general, because it may be easily edited, this is why I made such decision to read most important properties from the game data instead. Maybe that was another case of overthinking the problem, but I could not figure out what should engine/setup do if any of those fields are missing in config.

I would say that, since the config file shipped with the game has some options chosen by the game author, plus the defaults to get the engine to work, it should just be treated as part of the installation (i.e. it is a static component). If someone manually edits it and then the game doesn't start, that would be okay, they can just re-install it. Equally, if someone deletes the vox file and then says they have no sound, they can just re-install it. I'm not actually sure that a .cfg extension would default to opening in a text editor, so I think anyone savvy enough to be fiddling in there would also be savvy enough to fix what they break.

> Since we are at this, in the past we were discussing how could alternate (non-embedded) setup program get some data from the game or engine (for example, engine's version etc), and we thought that passing some data by standard output may be not a bad idea. For example, if engine is called with something like "--tell gfxdrivers" it will print the list of supported renderers that caller will receive, and so forth. I just thought I'd mention this, since this is also another option to get paths or config-related data from the game using engine as a tool to extract it.

To be honest, since the target of what I was making was non-Windows games uploaded to Gamejolt, Steam, or itch.io, and given that the engine already performs fallback actions when the chosen renderer is not available, I think it would probably be enough to just offer the basic settings for driver, get the screen geometry (just to work out which multipliers to offer) and let the engine deal with the rest. It will work on Windows too, and I think on Windows the engine has the same fallback behaviour when picking a renderer?

Even on Windows though, I get completely different results (and failures) just by using a second monitor, and after recently trying a 256 colour game for someone, my results didn't match with theirs (regardless of trying in a window or fullscreen, I couldn't get the  game working with anything but the software renderer).

Am I right in thinking that the config data is only going to be one other location per platform? If so, I could just hard-code it to look in certain folders, and put it in the main repo to keep it in sync with engine changes. This wasn't really meant to be a tool for launching a game from 10 years ago, but rather something to ship with a new game.

**Comment by ivan-mogilko** [comment]
> I would say that, since the config file shipped with the game has some options chosen by the game author, plus the defaults to get the engine to work, it should just be treated as part of the installation (i.e. it is a static component). If someone manually edits it and then the game doesn't start, that would be okay, they can just re-install it. Equally, if someone deletes the vox file and then says they have no sound, they can just re-install it.

Maybe I am looking at this from the different perspective, from engine's point of view currently it is not really a static component, none of the settings in config are cruicial for the game behavior, except maybe custom path to saves/appdata folder (when the game have some files preinstalled there, which is frankly not the way I hoped these will be used). I mean, 99% games still may be run without config, only difference will be that starting settings won't be the ones that author suggested.

In my opinion whether config should be treated as a static component is not a trivial question and the answer defines the rule for engine's behavior (when to run and when refuse to run, which options are obligatory and for games made with which version of AGS, because older games might not have it, and so on).
I don't know, maybe it's just how I look at things. Also over the past years particular game authors kept asking me to keep engine being able to run with defaults regardless of what happened to config file, and I tried to keep it that way (seems not perfectly so though #519).

We could make this game title (or user path) setting meant primarily for the setup program. But then engine must follow same rule, otherwise it may desync with setup program: it should read the setting from default config at start to know where the user config should be looked at. What happens if this setting is missing in cfg? If it's considered not mandatory, then engine could fallback to constructing path with the game title read from the game data.
But there's still a question what should a setup program do in same situation. If it's acceptable for your application to report error in such case (suggesting reinstall), then that solves this question.

If this option has to be mandatory though, then it radically changes the way engine treats cfg file, at least for the games made with AGS 3.5.0 and further.

For these reasons, and others, I still believe that interacting with the engine to get at least certain data may be a convenient option.


**EDIT:** By the way, this reminded me: currently the engine [enforces certain values when passing config to the setup program](https://github.com/adventuregamestudio/ags/blob/ags3/Engine/main/engine.cpp#L162) (which is possible because setup program is integrated into engine). For instance, it may disable "Render sprites at screen resolution" checkbox this way if it was disabled in game's General Settings. This data is passed in the exactly same form as everything else (key/value table read from INI), so meant to serve rather an additional "guarantee" than absolute necessity. Unfortunately, it seems that I forgot to write same instruction in the Editor... which should be fixed.


And of course if you think that config will be better at AppData/Local or Roaming on Windows, we may change it to there (users will be annoyed again though).


---


> I think it would probably be enough to just offer the basic settings for driver, get the screen geometry (just to work out which multipliers to offer) and let the engine deal with the rest. It will work on Windows too, and I think on Windows the engine has the same fallback behaviour when picking a renderer?

Yes, absolutely, all the setting behavior is uniform on all platforms, it may use slightly different defaults though.


> Am I right in thinking that the config data is only going to be one other location per platform?

There is certain difference between Windows and other platforms here. By default engine reads THREE configs: default (from game installation), global user config (meant for all games) and per-game user config (meant for one game). These configs are read in this order and override each other.
[This is explained in OPTIONS.md](https://github.com/adventuregamestudio/ags/blob/ags3/OPTIONS.md#configuration-file-locations).
Initially global config was also supported on Windows, but then someone asked me to disable it, being concerned that players may create/edit it by mistake and screw settings up for all the installed games unbeknownst to themselves. The biggest problem is that winsetup is not made for editing it.

Basically, this whole situation between config and setup GUI became "bit" messed up.


> If so, I could just hard-code it to look in certain folders, and put it in the main repo to keep it in sync with engine changes. This wasn't really meant to be a tool for launching a game from 10 years ago, but rather something to ship with a new game.

Do you mean Engine&Editor's repository? To be frank, I would not like to add additional tools there, because IMHO it is complicated enough. Besides these paths should change very rarely (if ever).

Or do you mean syncing some other behavior too?

**Comment by morganwillcock** [comment]
Thank you for the details, this does help to clarify things. I hadn't considered that at one point this was a three way merge, so I'll look into it further.

>Do you mean Engine&Editor's repository? To be frank, I would not like to add additional tools there, because IMHO it is complicated enough. Besides these paths should change very rarely (if ever).

>Or do you mean syncing some other behavior too?

If there was a way to access the game name, and the paths for where data may be rarely changes, then I think that just hardcoding the paths to check in the tool would work and nothing would really have to stay in sync. The situation only really came up because, I was testing a game with someone to upload to itch.io, and we ended up having to search the winsetup source code to find how the settings could be edited in order to fix the renderer performance (and after trying a few options and not managing to switch it to windowed mode, we game up).

I hadn't considered the 'render at screen resolution' lockout though, and I can see why some people would be quite passionate about not having their graphics exploded over a higher resolution.

For the moment I've spent most of the time trying to re-arrange and simplify the bits that were Windows specific (like the audio devices), so I've not invested too much time related to the actual settings. I guess the obvious solution would be to just removed anything but the most basic conrtols, since if you aren't using Windows you are probably only interested in the graphics driver and scaling options. I was also trying to broadly 'borrow' the layout of the Unity game launcher; I'll attach screenshots, although I hadn't entirely finished the layout.

![1](https://user-images.githubusercontent.com/14958747/47054000-7fe11080-d1a7-11e8-8143-31bd193bfe7f.png)
![2](https://user-images.githubusercontent.com/14958747/47054001-7fe11080-d1a7-11e8-9ffb-998d20b8afaf.png)

One thing that does seem to keep happening is that games get released without setting the 'titletext' value in the config file, so when you run winsetup it will say something like 'empty game setup'. It could be worth writing the game name into the setup file anyway, and not setting 'titletext' by default. This way it can default to $gamename + ' setup' for the winsetup title name, plus make the game name available for anything else.

**Comment by ivan-mogilko** [comment]
Alright, so, you need "game name" to be written to the default cfg. This may be added easily; but then engine should probably also use it if it's present to make sure it does not look in a different location for the user config.

I don't know why "titletext" is not written to config, it really should, at least if you are using recent Editor (3.4.1.15 and one in ags3 branch). Maybe worth to do a test.

> I hadn't considered the 'render at screen resolution' lockout though, and I can see why some people would be quite passionate about not having their graphics exploded over a higher resolution.

Yes, this was made so game author could lock this if they really want. Actually, engine does the lock too when necessary regardless of what is in config, but I've made winsetup to disable the checkbox to let players know they cannot change it.
There are other things that may get locked in winsetup, they are controlled by the "[disabled]" section in config. I forgot to document it in the OPTIONS.md, which I plan to fix very soon (also some things are outdated there).

**Comment by morganwillcock** [comment]
> EDIT: what may be wrong there is that TitleText property is only set once in the Editor and not updated when user changes game name. Perhaps it could remember whether user changed the TitleText itself, and if not then update automatically.

I think that it has to stay separate from the game name, since the window name could be just the game name, or it could a description of the setup program.

i.e. if my game was called "Bank Raid 2: It was a setup", I wouldn't want the window title to be automatically set to "Bank Raid 2: It was a setup setup"

**Comment by ivan-mogilko** [comment]
I remembered that knowing game name is not enough to get the savegame path. Savegame folder may have custom name (set in General Settings, "Save games folder name"). It is equal to game's title by default, but may be edited. (Not sure what was the original purpose for this, maybe to let combine saves from several minigames in one folder?)

Guess, either we need to write full necessary information to config then (both game name and savegame folder) or actually write config in a location unrelated to saves, like you suggested earlier.

**Comment by morganwillcock** [comment]
> Savegame folder may have custom name (set in General Settings, "Save games folder name")

Thanks. I'd not noticed this either, it is a lot more complicated than I had first realised.

I think I'd be tempted to leave everything as-is for version 3.5, and then look at this again for AGS 4 with the aim of using some sort of UUID for the game specific path (stored in %APPDATA% on Windows) and a 2 file settings merge (since on Windows the third config file is already disabled). I wouldn't want to try and force path identifiers into the config and potentially make it more complicated.

What is your general feeling on it?

Thank you for further investigation though, I much appreciate it.

**Comment by ivan-mogilko** [comment]
If we don't change anything in regards to how the path is constructed for now, the two possible options that come to mind are:
* make Editor create a path (or only the folder name) using same rules as the engine, and write it down in config as a hint for the setup programs;
* implement a cmd arg for the engine that make it return this path according to preloaded game data (e.g. something like "--tell savedir).

**Comment by morganwillcock** [comment]
For the second option, does the engine have to go through the graphics initialization in order to load data and return it? I wouldn't want to end up in a 'chicken and egg' situation, where to change options for the graphics renderer requires initializing the graphics renderer.

**Comment by ivan-mogilko** [comment]
> For the second option, does the engine have to go through the graphics initialization in order to load data and return it?

No, it is already taught to preload this data before anything else, because it needs it to run the integrated setup dialog on Windows. The initialization works roughly like this:

* parses command line;
* reads default config; this config may contain a direction to the main game data;
* finds main game data, if not found then quit;
* preloads the heading part (enough to get several needed variables);
* uses the game name and savegame path to find user config, reads user config;
* if command line told to run setup, then runs integrated setup dialog (only Windows version);
* loads main game data and checks for errors;
* if everything is ok, initializes graphics and display mode.

**Comment by morganwillcock** [comment]
Okay. Getting the engine to print game data sounds the better long term solution, as long as there is no ambiguity over delimiters or values with spaces in them.

**Comment by ericoporto** [comment]
The conclusion here is the engine should have `--tell` flags on the engine to get where the settings are written, where are the save files and what is the game name?

**Comment by sonneveld** [comment]
> The conclusion here is the engine should have `--tell` flags on the engine to get where the settings are written, where are the save files and what is the game name?

Yes! I think `json` would be a good format for this. It would make it easier to parse from the reader's perspective and allow us to extend the format easily.  We don't need to have a json writer initially, just a formatted string that translates to json would be a good start.

**Comment by ivan-mogilko** [comment]
I thought the original idea was to return single value by a requested key... e.g. `--tell configpath`

**Comment by morganwillcock** [comment]
Regardless of output format, if I wanted to know 10 config values then it would seem inefficient to run the program 10 times.

Regarding the output format, JSON is *probably* fine (I wanted to use the default TCL that comes with macOS and I don't know for sure that the JSON parser is in the core library) and makes more sense if the config data is anything more than a flat list. I'm not sure it is so easy to emit JSON manually where some of these values will be user defined, you'd probaly end up writing half a JSON encoder just to make sure the JSON was valid. For simplicitly I think I would prefer passing the delimiters and separators (if there was a namespace), and printing records this way.

**Comment by ericoporto** [comment]
I actually had the same thought Ivan had, I imagine there's no need at first to output any config file (which I imagine would be an ini file , like, `acsetup.cfg`) because the person who wants to read in bulk could just read that file. It would be useful, if that file is lost and one needs a clean one.

**Comment by ivan-mogilko** [comment]
> I imagine there's no need at first to output any config file (which I imagine would be an ini file , like, acsetup.cfg) because the person who wants to read in bulk could just read that file. It would be useful, if that file is lost and one needs a clean one.

Sorry, I am confused, what ini file? This was not about outputting a config file...

**Comment by ivan-mogilko** [comment]
> Regardless of output format, if I wanted to know 10 config values then it would seem inefficient to run the program 10 times.

So what is the suggestion then, return always everything engine can return, or support returning single value or a group of values? Maybe let user specify a list of keys too?

**Comment by ivan-mogilko** [comment]
Also, what's the bare minumum of returned data that is okay to support? I was thinking this may be added to 3.5.0 release if it's not something complex.

**Comment by ericoporto** [comment]
> Sorry, I am confused, what ini file? This was not about outputting a config file... or was it?

I thought the problem was just figuring the config file path and the game name. 

The output of a new acsetup.cfg was just me thinking what if the config file gets deleted for some reason, how to recreate it - for whatever other program that write/reads the config file to keep working. But I don't think this is needed for this issue.

**Comment by morganwillcock** [comment]
> The output of a new acsetup.cfg was just me thinking what if the config file gets deleted for some reason, how to recreate it - for whatever other program that write/reads the config file to keep working. But I don't think this is needed for this issue.

I think I follow you, that outputting all data from default and existing configurations in an ini format would give you a config file that was a merge of any existing settings.

> Also, what's the bare minumum of returned data that is okay to support? I was thinking this may be added to 3.5.0 release if it's not something complex.

For the three way merge, I think this will be tricky, and is probably where the above becomes relevant. If wanting to write new a new per game config file you would actually need the result of the merge in order for the initial display to accomodate the global and default files. This is assuming that the global file is given equal weight in terms of a configuration feature, and I'd probably suggest that this shouldn't be the case - especially if it is already disabled on the primary platform.

So if we say that defaults come from what ships with the game, and I am writing a copy of this, or editing an existing copy of this, I think it would need (from something like `--tell-settings`):
* game name
* path to the default config file (just in-case the filename ever changes in the future)
* path to the user configurable config file
* any values to lock or hide (i.e. render at screen resolution settings)

and then optionally (because the fallback behaviour might be safer, from something like `--tell-graphics`):
* available graphics drivers
* screen geometry

and then if you wanted all of the above in one go, you would just do `--tell-all`.

**Comment by ivan-mogilko** [comment]
Question: what is a better syntax to pass name of information you want to receive?
1. Having separate commands as in above:
`--tell-all`, `--tell-graphics` and so on;
2. Having a list of keywords in double quotes, like:
`--tell "graphics settings"`

**Comment by ericoporto** [comment]
2 appears to be more interesting, could the user still do `--tell all` ?

**Comment by morganwillcock** [comment]
I'd go for the first one, just to keep it simple and not require the calling environment to deal with the quoting (so usage will always look the same on all platforms).

**Comment by ivan-mogilko** [comment]
I did an experimental commit, only to try out the idea: https://github.com/ivan-mogilko/ags-refactoring/commit/60f5519ff49b61535f3ea36b7abf0e4bdf078352

So, it was decided to write everything in json format, correct?

**Comment by morganwillcock** [comment]
As long as the reserved characters are escaped, sure.
https://stackoverflow.com/a/27516892

Just printing things per line with a delimiter (e.g. a tab character or something user defined) also seems fine to me.

**Comment by ivan-mogilko** [comment]
Should I close this? #897 was merged.

**Comment by morganwillcock** [comment]
I think it is all covered now. Thanks for adding this.
